comments/serializers.py

```diff
@@ -66,6 +68,10 @@ def get_changed_my_mind(self, comment: Comment) -> dict[str, bool | int]:
     def get_text(self, value: Comment):
         return _("deleted") if value.is_soft_deleted else value.text
 
+    def get_edited_at(self, comment: Comment):
+        latest_diff = comment.comment_diffs.order_by('-created_at').first()
```

Comment:
note: I'm not much of a django dev. this does the right thing re: outputting a `select ... order by ... limit 1`, right? (rather than loading all the diffs into mem)

should we be preloading this or is it fine?

Reply:
Note also -- this seemed right vs using the `edited_at` field directly on the model because:

- the model edited_at is set for new unedited comments as well, for some reason
- the value that gets set is very slightly after the created_at field for whatever reason so you can't even do a `edited_at != created_at`
- the diff table just seems the more authoritative source for edits (I'm guessing the edited_at field might fire on a db trigger or something for internally-driven updates as well as user-initiated edits)

Reply:
This solution with produce [N+1 queries problem](https://docs.sentry.io/product/issues/issue-details/performance-issues/n-one-queries/).

> the value that gets set is very slightly after the created_at field for whatever reason so you can't even do a edited_at != created_at

I see what you mean. It’s likely that our `edited_at` field is automatically populated during question creation, but with a slight delay of a few microseconds from the created_at timestamp. This is expected since these operations execute sequentially.

To avoid unnecessary queries for such a simple task, I don’t think we need to fetch the edit history.

Here are two possible solutions:
- Simple workaround (my fav): Treat questions as “non-edited” if edited_at is either null or within 1 second of created_at.
- More complex solution: Refactor the logic for edited_at assignment and create a custom data migration to normalize the edited_at field for existing records.

Reply:
>This solution with produce [N+1 queries problem](https://docs.sentry.io/product/issues/issue-details/performance-issues/n-one-queries/).

Oh, duh, yep. Thanks for catching it.

>Simple workaround (my fav): Treat questions as “non-edited” if edited_at is either null or within 1 second of created_at.

I don't love this. It's pragmatic for sure, but it's a bit of a hack, no?

>More complex solution: Refactor the logic for edited_at assignment and create a custom data migration to normalize the edited_at field for existing records.

This feels like pulling on quite a thread. It looks like `edited_at` is used across many tables to mean the raw db updated time, so changing the definition/behaviour just for this table would be a bit inconsistent. One could introduce a new column for it, but ultimately it's still duplicating information vs the `comments_diff` table -- unless you want to get rid of it entirely in future?

How about option 3: https://docs.djangoproject.com/en/5.1/ref/models/querysets/#prefetch-related

This looks like it should avoid n+1, keep one source of truth re: edit history, and (if we want to) we can e.g. implement github style edit history UI for comments using it as well.

Ultimately though I'll take your steer as an official maintainer.

What do you think?

Reply:
cb4d47dd

I believe this should solve the n+1.

Reply:
`prefetch_related` won’t work when further filtering is applied to the prefetched queryset.

There are a couple of quick options:

- Prefetch with Python sorting & filtering — This approach avoids extra database queries:
  `sorted(comment.comment_diffs, key=lambda x: x.created_at, reverse=True)[0]`
  Cons: Prefetches the entire edit log, which may be more data than needed. Probably worth to make a
  `Prefetch("comment_diffs").only("created_at")` construction to select only created_at field from db
- Annotate aggregated `created_at` of the last CommentDiff object - this option adds an annotation for the `created_at` timestamp of the latest CommentDiff, for example:
  ```python
  comments.annotate(
      last_diff_created_at=SubqueryAggregate(
          "comment_diffs__created_at",
          aggregate=Max,
      )
  )
  ```
  Note: This may require [additional composite database indexes](https://docs.djangoproject.com/en/5.1/ref/models/options/#django.db.models.Options.indexes) for faster lookup on CommentDiff, such as:
  `Index("comment", "-created_at")`